### PR TITLE
4.x: Workspace Refactoring: Create workspace from template

### PR DESCRIPTION
**Pull request by deepakrkris**
Add loopback model to create workspace from templates

### CODE REVIEW

File: common/models/workspace.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+'use strict';
+
+/**
+  * Represents a LoopBack Workspace.
+  *
+  * @class Workspace
+  */
+module.exports = function(Workspace) {
+  Workspace.on('dataSourceAttached', function(eventData) {
+    const templateRegistry = require('../../component/template-registry');
```
Comment (superkhau) [inline common/models/workspace.js]: move import to top of file

File: server/server.js [hunk 2]
```diff
@@ -4,9 +4,16 @@
...
-var boot = require('loopback-boot');
-var app = module.exports = loopback();
+const loopback = require('loopback');
+const boot = require('loopback-boot');
+const app = module.exports = loopback();
+
+const templateRegistry = require('../component/template-registry');
+templateRegistry.loadTemplates(function(err, status) {
+  if (err) throw err;
+  console.log(status);
```
Comment (superkhau) [inline server/server.js]: remove

File: templates/api-server.json [hunk 3]
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "api-server",  
+  "description": "A LoopBack API server with local User auth",
+  "package": {
+    "version": "1.0.0",
+    "main": "server/server.js",
+    "scripts": {
+      "lint": "eslint .",
+      "start": "node .",
```
Comment (superkhau) [inline templates/api-server.json]: we should consider including `"test": "mocha"` in the default templates -- @bajtos thoughts? I feel it's weird how we can scaffold a LB app and not be able to run `npm test` OOB (since we're opinionated on other things like `eslint` anyways).

@deepakrkris This is outside the scope of this PR of course.

File: templates/api-server.json [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            "loopback/server/models",
+            "../common/models",
+            "./models"
+        ],
+        "mixins": [
+            "loopback/common/mixins",
+            "loopback/server/mixins",
+            "../common/mixins",
+            "./mixins"
+        ]
```
Comment (superkhau) [inline templates/api-server.json]: indentation is off

File: test/acceptance-tests/use-cases/test-scripts/create-from-templates.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+const Workspace = app.models.Workspace;
+app.on('booted', function() {
+  app.emit('ready');
+});
+
+module.exports = function() {
+  const testsuite = this;
+  this.Given(/^that the templates are loaded$/, function(next) {
+    testsuite.destinationPath =
+      path.resolve(__dirname, '../../../', 'sandbox2');
```
Comment (superkhau) [inline test/acceptance-tests/use-cases/test-scripts/create-from-templates.js]: I'm curious as to why we need sandbox2 -- can't we just wipe the original sandbox before executing your tests? (ie. empty out the original sandbox dir during the setup phase).

File: test/acceptance-tests/use-cases/test-scripts/create-from-templates.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    app.on('templates-loaded', function() {
```
Comment (superkhau) [inline test/acceptance-tests/use-cases/test-scripts/create-from-templates.js]: app.on('templates-loaded', next);

File: test/acceptance-tests/use-cases/test-scripts/create-from-templates.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      destinationPath: testsuite.destinationPath,
+    };
+    Workspace.create(data, {}, function(err, data) {
+      if (err) return next(err);
+      next();
+    });
+  });
+
+  this.Then(/^the workspace is created$/, function(next) {
+    next();
```
Comment (superkhau) [inline test/acceptance-tests/use-cases/test-scripts/create-from-templates.js]: missing assertions to verify the workspace has been created successfully

File: test/helpers/pretest.js [hunk 8]
```diff
@@ -1,6 +1,13 @@
...
+const sandboxDir2 = path.resolve(__dirname, '../sandbox2/');
 
-testSupport.givenEmptySandbox(function(err) {
+testSupport.givenEmptySandbox(sandboxDir, function(err) {
+  if (err) throw err;
+});
+
+testSupport.givenEmptySandbox(sandboxDir2, function(err) {
   if (err) throw err;
 });
```
Comment (superkhau) [inline test/helpers/pretest.js]: remove the duplication above

File: templates/api-server.json (hunk 3 above)
Comment (deepakrkris) [inline templates/api-server.json]: @bajtos @superkhau do you like the idea of having the template as a json file. the original workspace had it as a folder with .js files

**Comment by deepakrkris** [comment]
@slnode test please

File: templates/api-server.json (hunk 3 above)
Comment (superkhau) [inline templates/api-server.json]: See response at https://ibmapim.slack.com/archives/sq-lb-asteroid/p1483649471000026

File: templates/api-server.json (hunk 4 above)
Comment (deepakrkris) [inline templates/api-server.json]: agreed

File: test/acceptance-tests/use-cases/test-scripts/create-from-templates.js (hunk 5 above)
Comment (deepakrkris) [inline test/acceptance-tests/use-cases/test-scripts/create-from-templates.js]: agreed

File: test/acceptance-tests/use-cases/test-scripts/create-from-templates.js (hunk 6 above)
Comment (deepakrkris) [inline test/acceptance-tests/use-cases/test-scripts/create-from-templates.js]: I have added some logic before next now

File: test/acceptance-tests/use-cases/test-scripts/create-from-templates.js (hunk 7 above)
Comment (deepakrkris) [inline test/acceptance-tests/use-cases/test-scripts/create-from-templates.js]: agreed

File: test/helpers/pretest.js (hunk 8 above)
Comment (deepakrkris) [inline test/helpers/pretest.js]: agreed

File: server/server.js [hunk 9]
```diff
@@ -4,9 +4,15 @@
 // License text available at https://opensource.org/licenses/MIT
 'use strict';
 
-var loopback = require('loopback');
-var boot = require('loopback-boot');
-var app = module.exports = loopback();
+const loopback = require('loopback');
+const boot = require('loopback-boot');
+const app = module.exports = loopback();
```
Comment (superkhau) [inline server/server.js]: //imports

module.exports = app;

const app = loopback();

File: test/acceptance-tests/use-cases/test-scripts/create-from-templates.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+app.on('booted', function() {
+  app.emit('ready');
+});
+
+module.exports = function() {
+  const testsuite = this;
+  this.Given(/^that the templates are loaded$/, function(next) {
+    testsuite.destinationPath = testSupport.givenSandboxDir();
+    app.on('templates-loaded', function() {
+      console.log('templates-loaded');
```
Comment (superkhau) [inline test/acceptance-tests/use-cases/test-scripts/create-from-templates.js]: delete

File: test/acceptance-tests/use-cases/test-scripts/create-from-templates.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+const Workspace = app.models.Workspace;
+app.on('booted', function() {
+  app.emit('ready');
+});
+
+module.exports = function() {
+  const testsuite = this;
+  this.Given(/^that the templates are loaded$/, function(next) {
+    testsuite.destinationPath = testSupport.givenSandboxDir();
+    app.on('templates-loaded', function() {
```
Comment (superkhau) [inline test/acceptance-tests/use-cases/test-scripts/create-from-templates.js]: app.on('templates-loaded', next);

File: test/acceptance-tests/use-cases/test-scripts/create-from-templates.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+  });
+
+  this.When(/^I create a workspace from the template '(.+)'$/,
+  function(templateName, next) {
+    testsuite.templateName = templateName;
+    const data = {
+      templateName: testsuite.templateName,
+      destinationPath: testsuite.destinationPath,
+    };
+    Workspace.create(data, {}, function(err, data) {
```
Comment (superkhau) [inline test/acceptance-tests/use-cases/test-scripts/create-from-templates.js]: pass next

**Comment by deepakrkris** [comment]
@superkhau will merge after CI is up and running

**Comment by deepakrkris** [comment]
@slnode test please

**Comment by rmg** [comment]
@slnode test please

**Comment by rmg** [comment]
@slnode test please

**Comment by rmg** [comment]
FYI, the **loopback-workspace.dependants** job no longer exists, so the ❌  beside it is misleading. Rebasing or refreshing the branch will clear it.

File: templates/api-server.json (hunk 3 above)
Comment (bajtos) [inline templates/api-server.json]: @superkhau @deepakrkris could you please summarise the outcome of the discussion on Slack, so that people reading this patch can access it too, even if they don't have access to our Slack?

Also I see that this pull request was already merged. If we are going to continue this discussion, then please open a new GH issue where we can leave our comments.

> do you like the idea of having the template as a json file. the original workspace had it as a folder with .js files

This is not only about JSON vs JS extension. The more important question is whether we want the templates to contain content as it will be scaffolded (i.e. `server/middleware.json` in the current template), or whether we want templates to contain values for Domain Entities provided by workspace?

The former has the benefit that templates are easy to maintain and it's immediately clear what will be written to the filesystem.

The latter has the benefit that a single template can support different configuration formats. For example, we have changed the way how remote methods are defined in LoopBack 3.0. At the same time, we improved loopback-workspace to support both flavours in a single version. That way a single `ModelMethod` template can be used to scaffold the relevant entry in `common/models/my-model.json` regardless of what LoopBack version the app is using. If we were using the first approach, then we would have to maintain multiple versions of the json template (LB 1-2.x, LB 3.x+).

Comment (bajtos) [inline templates/api-server.json]: > we should consider including "test": "mocha" in the default templates -- @bajtos thoughts?

@superkhau That would make sense to me only if we included some sample (smoke?) tests in the template too. If you think this is worth pursuing, then please open a new GH issue to discuss.

Comment (superkhau) [inline templates/api-server.json]: >That would make sense to me only if we included some sample (smoke?) tests in the template too. If you think this is worth pursuing, then please open a new GH issue to discuss.

@bajtos I was thinking along the same lines (add smoke tests in the templates too). Whether the time is now is up for debate. 

@deepakrkris Can you add a summary of the Slack discussion to the bottom of this PR and your reasoning as for landing this now and what you intend to do in the future so people reading this can follow up? Maybe open up a new issue with a discussion label to gather some community feedback on the json vs js notes and @bajtos's input from above together?
